security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -191,35 +144,136 @@ func (s *sdsservice) generate(resourceNames []string) (model.Resources, error) {
 	return resources, nil
 }
 
-// Generate implements the XDS Generator interface. This allows the XDS server to dispatch requests
-// for SecretTypeV3 to our server to generate the Envoy response.
-func (s *sdsservice) Generate(proxy *model.Proxy, w *model.WatchedResource, updates *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
-	// updates.Full indicates we should do a complete push of all updated resources
-	// In practice, all pushes should be incremental (ie, if the `default` cert changes we won't push
-	// all file certs).
-	if updates.Full {
-		resp, err := s.generate(w.ResourceNames)
-		return resp, pushLog(w.ResourceNames), err
+// register adds the SDS handle to the grpc server
+func (s *sdsservice) register(rpcs *grpc.Server) {
+	sds.RegisterSecretDiscoveryServiceServer(rpcs, s)
+}
+
+func (s *sdsservice) push(name string) {
+	s.Lock()
+	defer s.Unlock()
+	for _, client := range s.clients {
+		go func() {
+			select {
+			case client.Connection().PushChannel <- name:
+			case <-client.Connection().StreamClosed():
+			}
+		}()
+	}
+}
+
+func (c Context) Connection() *xds.BaseConnection {
+	return &c.BaseConnection
+}
+
+var connectionNumber = int64(0)
+
+func (c *Context) Initialize(node *core.Node) error {
+	id := atomic.AddInt64(&connectionNumber, 1)
+	con := c.Connection()
+	con.SetID(strconv.FormatInt(id, 10))
+
+	c.s.Lock()
+	c.s.clients[con.ID()] = c
+	c.s.Unlock()
+
+	con.DoneInitialize()
+	return nil
+}
+func (c *Context) Close() {
+	c.s.Lock()
+	defer c.s.Unlock()
+	delete(c.s.clients, c.Connection().ID())
+}
+func (c *Context) Watcher() xds.Watcher {
+	return c.w
+}
+func (c *Watch) DeleteWatchedResource(string) {
+	c.Lock()
+	defer c.Unlock()
+	c.watch = nil
+}
+func (c *Watch) GetWatchedResource(string) *xds.WatchedResource {
+	c.Lock()
+	defer c.Unlock()
+	return c.watch
+}
+func (c *Watch) NewWatchedResource(typeURL string, names []string) {
+	c.Lock()
+	defer c.Unlock()
+	c.watch = &xds.WatchedResource{TypeUrl: typeURL, ResourceNames: names}
+}
+func (c *Watch) UpdateWatchedResource(_ string, f func(*xds.WatchedResource) *xds.WatchedResource) {
+	c.Lock()
+	defer c.Unlock()
+	c.watch = f(c.watch)
+}
+func (c *Watch) GetID() string {
+	return ""
```

Comment:
can we return a useful ID here? Maybe we can use the fact we know we only subscribe to 1 resource to make the resource name the log, which would be super handy. Although for the initial connect it would be not yet available

Reply:
I'm worried that grabbing a lock can cause issues so the list of resources is probably not easy. Not sure what to put here.